update etcd, grpc, protobuf dependencies

liggitt opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Fixes https://github.com/kubernetes/kubernetes/issues/93320

Updates to a version of github.com/golang/protobuf that resolves immediate gogo protobuf incompatibilities as a short-term fix (https://go-review.googlesource.com/c/protobuf/+/300869/).

Update protobuf,grpc,etcd dependencies

1. Updated etcd/protobuf/grpc dependencies:

    ```
    echo "
    hack/pin-dependency.sh github.com/golang/protobuf latest
    hack/pin-dependency.sh google.golang.org/protobuf latest
    hack/pin-dependency.sh go.etcd.io/etcd/api/v3 v3.5.0-rc.0
    hack/pin-dependency.sh go.etcd.io/etcd/client/v3 v3.5.0-rc.0
    hack/pin-dependency.sh go.etcd.io/etcd/client/pkg/v3 v3.5.0-rc.0
    hack/pin-dependency.sh go.etcd.io/etcd/pkg/v3 v3.5.0-rc.0
    hack/pin-dependency.sh go.etcd.io/etcd/server/v3 v3.5.0-rc.0
    hack/pin-dependency.sh go.etcd.io/etcd/tests/v3 v3.5.0-rc.0
    hack/pin-dependency.sh google.golang.org/grpc latest
    " | bash
    ```

2. Linted transitive dependencies until versions are clean:

    ```
    hack/lint-dependencies.sh  | grep "  hack/pin-dependency.sh" | bash
    ```

3. Linted dependencies until dropped versions are clean:

    ```
    hack/lint-dependencies.sh  | grep "dropreplace" | bash
    ```

4. Updated vendor and internal modules:

    ```
    hack/update-vendor.sh
    hack/update-internal-modules.sh
    ```

Repeated steps 2-4 until clean

Commit reviews by area:

- storage (@msau42 - https://github.com/kubernetes/kubernetes/pull/100488#issuecomment-849078167)
  - [x] Fix CSI mock driver to get marshaleable grpc error
- api-machinery / etcd (@ptabor - https://github.com/kubernetes/kubernetes/pull/100488#pullrequestreview-684054336)
  - [x] Drop etcd v2 support from etcd migration image
  - [x] Drop etcd v2 support from test util
  - [x] Switch to non-deprecated timestamppb.Now()
  - [x] Switch to go.etcd.io/etcd/client/v3
  - [x] Drop use of deprecated clientv3.SetLogger
  - [x] Quiet embedded etcd logs
  - [x] Eliminate parallel and unnecessary embedded etcd instances (best reviewed ignoring whitespace changes)
- api-machinery/network proxy (@Jefftree)
  - [x] Exercise egress configuration in local-up-cluster
  - [x] Fix etcd egress dialer addr parsing
- code-organization (@dims - https://github.com/kubernetes/kubernetes/pull/100488#issuecomment-861514855)
  - [x] Update protobuf,grpc,etcd dependencies

before merge:
- [x] wait for 1.22
- [x] get review of CSI mock driver change by sig-storage - https://github.com/kubernetes/kubernetes/pull/100488#issuecomment-849078167
- [x] wait for bbolt 1.3.6 (https://github.com/etcd-io/bbolt/issues/278)

post-merge:
- [x] update to etcd v3.5.0+ release - https://github.com/kubernetes/kubernetes/pull/102897
- [x] switch use of integration server to direct use of embedded server - https://github.com/kubernetes/kubernetes/pull/102883

/milestone v1.22
/cc @dims


```release-note
The etcd client has been updated to 3.5.0; github.com/golang/protobuf, google.golang.org/protobuf, and google.golang.org/grpc have been updated to current versions.
```

liggitt committed:
Drop etcd v2 support from etcd migration image

liggitt committed:
Drop etcd v2 support from test util

liggitt committed:
Switch to go.etcd.io/etcd/client/v3

fedebongio commented:
/triage accepted

liggitt committed:
Switch to non-deprecated timestamppb.Now()

liggitt commented:
seeing repeated CSI failures across pull-kubernetes-e2e-kind and pull-kubernetes-e2e-gce-ubuntu-containerd:

```
Kubernetes e2e suite: [sig-storage] CSI mock volume storage capacity exhausted, late binding, with topology

/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/csi_mock_volume.go:958
Mar 23 16:26:33.553: while waiting for all CSI calls
Unexpected error:
    <*errors.errorString | 0xc002804dc0>: {
        s: "error waiting for expected CSI calls: Unexpected CSI call 0: expected CreateVolume (8), got CreateVolume (0)",
    }
    error waiting for expected CSI calls: Unexpected CSI call 0: expected CreateVolume (8), got CreateVolume (0)
occurred
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/csi_mock_volume.go:1050

Kubernetes e2e suite: [sig-storage] CSI mock volume storage capacity exhausted, immediate binding

/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/csi_mock_volume.go:958
Mar 23 16:37:54.612: while waiting for all CSI calls
Unexpected error:
    <*errors.errorString | 0xc0027ddc30>: {
        s: "error waiting for expected CSI calls: Unexpected CSI call 0: expected CreateVolume (8), got CreateVolume (0)",
    }
    error waiting for expected CSI calls: Unexpected CSI call 0: expected CreateVolume (8), got CreateVolume (0)
occurred
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/csi_mock_volume.go:1050

Kubernetes e2e suite: [sig-storage] CSI mock volume storage capacity exhausted, late binding, no topology

/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/csi_mock_volume.go:958
Mar 23 16:22:44.756: while waiting for all CSI calls
Unexpected error:
    <*errors.errorString | 0xc003ad3720>: {
        s: "error waiting for expected CSI calls: Unexpected CSI call 0: expected CreateVolume (8), got CreateVolume (0)",
    }
    error waiting for expected CSI calls: Unexpected CSI call 0: expected CreateVolume (8), got CreateVolume (0)
occurred
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/storage/csi_mock_volume.go:1050
```

that seems likely to be related as those are proto APIs

liggitt committed:
Fix CSI mock driver to get marshaleable grpc error

ptabor commented:
From etcd experience, the binaries that depend on gogo (even transitively) can go only as far as: 
  - google.golang.org/grpc v1.32.0
  - github.com/gogo/protobuf v1.3.2
  - github.com/golang/protobuf v1.3.5
Further upgrades would require gogo / protobuf-runtime changes.

liggitt commented:
> Further upgrades would require gogo / protobuf-runtime changes

Are those not the changes in https://go-review.googlesource.com/c/protobuf/+/300869/ / github.com/golang/protobuf@v1.5.1 / google.golang.org/protobuf@v1.26.0 ?

ptabor commented:
> > Further upgrades would require gogo / protobuf-runtime changes
> 
> Are those not the changes in https://go-review.googlesource.com/c/protobuf/+/300869/ / github.com/golang/protobuf@v1.5.1 / google.golang.org/protobuf@v1.26.0 ?

Thank you for the link. I was looking forward to these changes, but didn't spotted them in the changelog. 
Really nice. I will test them with etcd.

ptabor commented:
Confirmed that it works nicely for etcd:   https://github.com/etcd-io/etcd/pull/12801 and the PR is merged. 
So etcd-3.5 is on the way of using grpc-1.36 & proto 1.5.1.

liggitt commented:
/skip

liggitt commented:
/retest

liggitt commented:
/test pull-kubernetes-e2e-gce-csi-serial

liggitt commented:
/hold for 1.22, etcd 3.5

liggitt committed:
Drop use of deprecated clientv3.SetLogger

liggitt commented:
not sure what is going on with the `pull-kubernetes-e2e-gce-storage-snapshot` failure:

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100488/pull-kubernetes-e2e-gce-storage-snapshot/1394693877893435392/

repeated many times:

> WARNING: pod log: csi-gce-pd-node-8d6dm/gce-pd-driver: the server does not allow this method on the requested resource ( pods/log csi-gce-pd-node-8d6dm)

PreStop hook errors:

> May 18 17:23:51.161: INFO: At 2021-05-18 17:23:50 +0000 UTC - event for csi-gce-pd-node-ncz2q: {kubelet e2e-0cd474acb7-30ee2-minion-group-fwc2} FailedPreStopHook: Exec lifecycle hook ([/bin/sh -c rm -rf /registration/pd.csi.storage.gke.io /registration/pd.csi.storage.gke.io-reg.sock]) for Container "csi-driver-registrar" in Pod "csi-gce-pd-node-ncz2q_snapshotting-391-7646(71591837-9a5e-479e-9d4f-6628e44897f0)" failed - error: rpc error: code = Unknown desc = failed to exec in container: failed to start exec "1e318c52b4dfaf74266b91610523831c4c99c4ffc0de908fc832855227289a4c": OCI runtime exec failed: exec failed: container_linux.go:349: starting container process caused "exec: \"/bin/sh\": stat /bin/sh: no such file or directory": unknown, message: ""

liggitt commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel
/test pull-kubernetes-e2e-gce-ubuntu-containerd

liggitt commented:
/sig storage
The CSI and storage snapshot jobs are unhappy with the grpc/protobuf updates. Can someone from the sig dig into those two job failures?

I already applied a fix to the mock CSI driver to get the error unmarshaled correctly, since `vendor/google.golang.org/grpc/status/status.go` changed errors in ways that don't marshal to json the same way (see the `Fix CSI mock driver to get marshaleable grpc error` commit). That fixed the `error waiting for expected CSI calls: Unexpected CSI call` errors seen in [blocking e2e runs](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100488/pull-kubernetes-e2e-gce-ubuntu-containerd/1374587601821896704).

/sig node
The node-e2e-containerd job is unhappy with the grpc/protobuf updates. Can someone from the sig dig into that job failure?

SergeyKanzhelev commented:
/cc

dims commented:
/test pull-kubernetes-node-e2e

msau42 commented:
pd csi e2e test failures are being fixed in https://github.com/kubernetes/kubernetes/pull/102212

SergeyKanzhelev commented:
> /sig node
> The node-e2e-containerd job is unhappy with the grpc/protobuf updates. Can someone from the sig dig into that job failure?

Test is failing with [build log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/100488/pull-kubernetes-node-e2e-containerd/1397577479983468544/build-log.txt) failure:

```
failed to start internal services: e2e service \"services\" readiness check timeout 2m0s
```

[Services.log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/100488/pull-kubernetes-node-e2e-containerd/1397577479983468544/artifacts/tmp-node-e2e-d48a6649-cos-81-12871-1290-2/services.log) ends with the call:

```
I0526 15:48:45.849561    1101 internal_services.go:58] Starting e2e services...
I0526 15:48:45.849568    1101 internal_services.go:112] Starting etcd
```

Fishing for the last successful execution from the [build history](https://prow.k8s.io/pr-history/?org=kubernetes&repo=kubernetes&pr=100488), [services.log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/100488/pull-kubernetes-node-e2e-containerd/1394693873627828224/artifacts/tmp-node-e2e-e6aea466-cos-81-12871-1290-2/services.log) pass forward to:

```
I0518 16:50:41.639310    1294 internal_services.go:58] Starting e2e services...
I0518 16:50:41.639319    1294 internal_services.go:112] Starting etcd
I0518 16:50:41.755667    1294 internal_services.go:121] Starting API server
I0518 16:50:41.755793    1294 plugins.go:84] "Registered admission plugin" plugin="NamespaceLifecycle"
...
```

Which indicates that the call to start Etcd in this file `test/e2e_node/services/internal_services.go` failed.

I'm not sure where the error is going to... I will get back to it later, posting here in case somebody want to take a look.

liggitt commented:
/test pull-kubernetes-e2e-gce-storage-snapshot
/test pull-kubernetes-e2e-gce-csi-serial

odinuge commented:
As a follow up from @SergeyKanzhelev, etcd fails to start because of:

```
Found leaked goroutined BEFORE test: appears to have leaked :                                                                                                                                                                                                                                                                   
k8s.io/kubernetes/vendor/go.opencensus.io/stats/view.(*worker).start(0xc000488140)
        /home/odin/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/go.opencensus.io/stats/view/worker.go:154 +0xcd
created by k8s.io/kubernetes/vendor/go.opencensus.io/stats/view.init.0
        /home/odin/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/go.opencensus.io/stats/view/worker.go:32 +0x57

k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).flushDaemon(0x8b87dc0)
        /home/odin/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1169 +0x8b
created by k8s.io/kubernetes/vendor/k8s.io/klog/v2.init.0
        /home/odin/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:420 +0xdf
```

Pretty sure it is because of https://github.com/etcd-io/etcd/pull/12753, that checks this before running the tests. Since the services things are wrapped in a go test, something fishy is going on, and this is not printed. It is just sent to `test.Skip` here: 

https://github.com/etcd-io/etcd/blob/e2d67f2e3bfa6f72178e26557bb22cc1482c418c/client/pkg/testutil/leak.go#L101

Did a _very_ quick look in `go.opencensus.io`, and it looks like there isn't a simple way to disable that worker. I have no idea why it is in the binary in the first place tho. Looks like klog has to be added to the exceptions list, and probably opencensus as well?

ptabor commented:
Interesting: 

We didn't consider (had awareness) etcd integration tests package as public API that can be used by other projects. 

https://github.com/kubernetes/kubernetes/blob/29e5e4df4b150169ed27f2c22e026800da9761cf/staging/src/k8s.io/apiserver/pkg/storage/etcd3/testing/test_server.go#L293

It's rather dirty and frequently changed. This motivates exposing well sanitized external variant of the package.

odinuge commented:
> We didn't consider (had awareness) etcd integration tests package as public API that can be used by other projects.

Yeah, you found the correct entry point in k8s. When the test fail/deamon doesn't work, it is more or less impossible to see it without digging deep into it and view how it works. I do agree that it is quite strange, and it took me quite some time to realize that the code was in "testing mode", when we just use it to spin up a set of daemons in a cli command. :sweat_smile: 

> It's rather dirty and frequently changed. This motivates exposing well sanitized external variant of the package.

Yeah, thats a good point. I guess we could also start etcd in another way, but I have not looked at why we use this functionality in the first place..

ptabor commented:
I would recommend going into direction of changing: 

```
func BeforeTest(t testutil.TB) { ... } in `integration/testing.go:39`
```

into 

```
func BeforeTest(t testutil.TB, opts ...BeforeTestOptions) { ... }
```

such that k8s can call: 
```
 integration.BeforeTest(t, integration.WithoutLeakDetection())
```

liggitt commented:
thanks for digging, opened https://github.com/etcd-io/etcd/issues/13051 with the three items I found while trying to work around the issue

liggitt commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

ptabor commented:
https://github.com/etcd-io/etcd/commit/6cd7c02873e2fe3ffe75627ca9d1ae0f953dce57 contains the (merged into release-3.5) fixes to the `integration` package. Should be incorporated in next -prerelease this week.

liggitt commented:
updated with etcd 3.5.0-rc.0

liggitt commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

liggitt committed:
Eliminate parallel and unnecessary embedded etcd instances

liggitt committed:
Quiet embedded etcd logs

liggitt commented:
fixed some parallel embedded etcd runs in unit tests and quieted embedded etcd logs

there were two failed packages that detected races in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100488/pull-kubernetes-unit/1400643717106241536 without any additional details in the log I want to keep an eye on:
```
=== CONT  
    testing.go:1092: race detected during execution of test
FAIL
FAIL	k8s.io/kubernetes/pkg/controlplane	94.608s
```

```
=== RUN   TestServiceUpgrade
=== CONT  TestServiceUpgrade
    testing.go:1092: race detected during execution of test
--- FAIL: TestServiceUpgrade (5.39s)
    logger.go:130: 2021-06-04T03:01:10.028Z	INFO	grpc	[[balancer] base.baseBalancer: handle SubConn state change: 0xc00188e510, CONNECTING]
```

liggitt commented:
/skip
/test pull-kubernetes-e2e-gce-csi-serial

liggitt commented:
/skip

liggitt commented:
/test pull-kubernetes-e2e-gce-csi-serial

liggitt commented:
/clear

BenTheElder commented:
needs-rebase (go.mod, go.sum, vendor/modules.txt)
/needs-rebase

liggitt commented:
looks like some tests are failing with etcd storage errors:

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/100488/pull-kubernetes-e2e-gce-storage-snapshot/1404458253030526976/artifacts/e2e-5eb2956713-30ee2-master/kube-apiserver.log

```
I0614 15:42:55.988465       9 storage_factory.go:285] storing apiServerIPInfo in v1, reading as __internal from storagebackend.Config{Type:"etcd3", Prefix:"/registry", Transport:storagebackend.TransportConfig{ServerList:[]string{"https://127.0.0.1:2379"}, KeyFile:"/etc/srv/kubernetes/pki/etcd-apiserver-client.key", CertFile:"/etc/srv/kubernetes/pki/etcd-apiserver-client.crt", TrustedCAFile:"/etc/srv/kubernetes/pki/etcd-apiserver-ca.crt", EgressLookup:(egressselector.Lookup)(0x3d93ac0)}, Paging:true, Codec:runtime.Codec(nil), EncodeVersioner:runtime.GroupVersioner(nil), Transformer:value.Transformer(nil), CompactionInterval:300000000000, CountMetricPollPeriod:60000000000, DBMetricPollInterval:30000000000, HealthcheckTimeout:2000000000, LeaseManagerConfig:etcd3.LeaseManagerConfig{ReuseDurationSeconds:60, MaxObjectCount:1000}}
I0614 15:42:55.989055       9 clientconn.go:252] [core] parsed scheme: "etcd-endpoints"
I0614 15:42:55.989099       9 resolver_conn_wrapper.go:96] [core] ccResolverWrapper: sending update to cc: {[{127.0.0.1:2379 127.0.0.1 <nil> 0 <nil>}] 0xc000b48300 <nil>}
I0614 15:42:55.989111       9 clientconn.go:708] [core] ClientConn switching balancer to "round_robin"
I0614 15:42:55.989119       9 clientconn.go:723] [core] Channel switches to new LB policy "round_robin"
I0614 15:42:55.989138       9 balancer.go:103] [balancer] base.baseBalancer: got new ClientConn state:  {{[{127.0.0.1:2379 127.0.0.1 <nil> 0 <nil>}] 0xc000b48300 <nil>} <nil>}
I0614 15:42:55.989158       9 clientconn.go:1114] [core] Subchannel Connectivity change to CONNECTING
I0614 15:42:55.989178       9 clientconn.go:1251] [core] Subchannel picks a new address "127.0.0.1:2379" to connect
W0614 15:42:55.989219       9 clientconn.go:1326] [core] grpc: addrConn.createTransport failed to connect to {127.0.0.1:2379 127.0.0.1 <nil> 0 <nil>}. Err: connection error: desc = "transport: Error while dialing parse \"127.0.0.1:2379\": first path segment in URL cannot contain colon". Reconnecting...
I0614 15:42:55.989230       9 clientconn.go:1114] [core] Subchannel Connectivity change to TRANSIENT_FAILURE
I0614 15:42:55.990456       9 balancer.go:207] [balancer] base.baseBalancer: handle SubConn state change: 0xc001306d20, CONNECTING
I0614 15:42:55.990488       9 clientconn.go:435] [core] Channel Connectivity change to CONNECTING
I0614 15:42:55.990497       9 balancer.go:207] [balancer] base.baseBalancer: handle SubConn state change: 0xc001306d20, TRANSIENT_FAILURE
I0614 15:42:55.990511       9 clientconn.go:435] [core] Channel Connectivity change to TRANSIENT_FAILURE
I0614 15:42:55.993423       9 clientconn.go:1114] [core] Subchannel Connectivity change to READY
I0614 15:42:55.993447       9 balancer.go:207] [balancer] base.baseBalancer: handle SubConn state change: 0xc000f925b0, READY
I0614 15:42:55.993480       9 roundrobin.go:50] [roundrobin] roundrobinPicker: newPicker called with info: {map[0xc000f925b0:{{127.0.0.1:2379 127.0.0.1 <nil> 0 <nil>}}]}
I0614 15:42:55.993492       9 clientconn.go:435] [core] Channel Connectivity change to READY
I0614 15:42:55.994683       9 clientconn.go:435] [core] Channel Connectivity change to SHUTDOWN
I0614 15:42:55.994703       9 clientconn.go:1114] [core] Subchannel Connectivity change to SHUTDOWN
I0614 15:42:55.994764       9 controlbuf.go:521] [transport] transport: loopyWriter.run returning. connection error: desc = "transport is closing"
I0614 15:42:56.582415       9 clientconn.go:1114] [core] Subchannel Connectivity change to CONNECTING
I0614 15:42:56.582460       9 clientconn.go:1251] [core] Subchannel picks a new address "127.0.0.1:2379" to connect
I0614 15:42:56.582634       9 balancer.go:207] [balancer] base.baseBalancer: handle SubConn state change: 0xc000b15090, CONNECTING
I0614 15:42:56.585955       9 clientconn.go:1114] [core] Subchannel Connectivity change to READY
I0614 15:42:56.585982       9 balancer.go:207] [balancer] base.baseBalancer: handle SubConn state change: 0xc000b15090, READY
I0614 15:42:56.586019       9 roundrobin.go:50] [roundrobin] roundrobinPicker: newPicker called with info: {map[0xc000b15090:{{127.0.0.1:2379 127.0.0.1 <nil> 0 <nil>}}]}
I0614 15:42:56.586029       9 clientconn.go:435] [core] Channel Connectivity change to READY
I0614 15:42:56.990133       9 clientconn.go:1114] [core] Subchannel Connectivity change to CONNECTING
I0614 15:42:56.990175       9 clientconn.go:1251] [core] Subchannel picks a new address "127.0.0.1:2379" to connect
W0614 15:42:56.990218       9 clientconn.go:1326] [core] grpc: addrConn.createTransport failed to connect to {127.0.0.1:2379 127.0.0.1 <nil> 0 <nil>}. Err: connection error: desc = "transport: Error while dialing parse \"127.0.0.1:2379\": first path segment in URL cannot contain colon". Reconnecting...
I0614 15:42:56.990227       9 clientconn.go:1114] [core] Subchannel Connectivity change to TRANSIENT_FAILURE
I0614 15:42:56.990244       9 balancer.go:207] [balancer] base.baseBalancer: handle SubConn state change: 0xc001306d20, CONNECTING
I0614 15:42:56.990277       9 balancer.go:207] [balancer] base.baseBalancer: handle SubConn state change: 0xc001306d20, TRANSIENT_FAILURE
I0614 15:42:58.666299       9 clientconn.go:1114] [core] Subchannel Connectivity change to CONNECTING
I0614 15:42:58.666350       9 clientconn.go:1251] [core] Subchannel picks a new address "127.0.0.1:2379" to connect
W0614 15:42:58.666405       9 clientconn.go:1326] [core] grpc: addrConn.createTransport failed to connect to {127.0.0.1:2379 127.0.0.1 <nil> 0 <nil>}. Err: connection error: desc = "transport: Error while dialing parse \"127.0.0.1:2379\": first path segment in URL cannot contain colon". Reconnecting...
I0614 15:42:58.666416       9 clientconn.go:1114] [core] Subchannel Connectivity change to TRANSIENT_FAILURE
I0614 15:42:58.666436       9 balancer.go:207] [balancer] base.baseBalancer: handle SubConn state change: 0xc001306d20, CONNECTING
I0614 15:42:58.666446       9 balancer.go:207] [balancer] base.baseBalancer: handle SubConn state change: 0xc001306d20, TRANSIENT_FAILURE
I0614 15:43:00.830930       9 clientconn.go:1114] [core] Subchannel Connectivity change to CONNECTING
I0614 15:43:00.830980       9 clientconn.go:1251] [core] Subchannel picks a new address "127.0.0.1:2379" to connect
W0614 15:43:00.831124       9 clientconn.go:1326] [core] grpc: addrConn.createTransport failed to connect to {127.0.0.1:2379 127.0.0.1 <nil> 0 <nil>}. Err: connection error: desc = "transport: Error while dialing parse \"127.0.0.1:2379\": first path segment in URL cannot contain colon". Reconnecting...
I0614 15:43:00.831154       9 clientconn.go:1114] [core] Subchannel Connectivity change to TRANSIENT_FAILURE
I0614 15:43:00.831173       9 balancer.go:207] [balancer] base.baseBalancer: handle SubConn state change: 0xc001306d20, CONNECTING
I0614 15:43:00.831182       9 balancer.go:207] [balancer] base.baseBalancer: handle SubConn state change: 0xc001306d20, TRANSIENT_FAILURE
I0614 15:43:04.126727       9 clientconn.go:1114] [core] Subchannel Connectivity change to CONNECTING
I0614 15:43:04.126768       9 clientconn.go:1251] [core] Subchannel picks a new address "127.0.0.1:2379" to connect
W0614 15:43:04.126818       9 clientconn.go:1326] [core] grpc: addrConn.createTransport failed to connect to {127.0.0.1:2379 127.0.0.1 <nil> 0 <nil>}. Err: connection error: desc = "transport: Error while dialing parse \"127.0.0.1:2379\": first path segment in URL cannot contain colon". Reconnecting...
I0614 15:43:04.126844       9 clientconn.go:1114] [core] Subchannel Connectivity change to TRANSIENT_FAILURE
I0614 15:43:04.126861       9 balancer.go:207] [balancer] base.baseBalancer: handle SubConn state change: 0xc001306d20, CONNECTING
I0614 15:43:04.126870       9 balancer.go:207] [balancer] base.baseBalancer: handle SubConn state change: 0xc001306d20, TRANSIENT_FAILURE
I0614 15:43:11.928403       9 clientconn.go:1114] [core] Subchannel Connectivity change to CONNECTING
I0614 15:43:11.928482       9 clientconn.go:1251] [core] Subchannel picks a new address "127.0.0.1:2379" to connect
I0614 15:43:11.928573       9 balancer.go:207] [balancer] base.baseBalancer: handle SubConn state change: 0xc001306d20, CONNECTING
W0614 15:43:11.928639       9 clientconn.go:1326] [core] grpc: addrConn.createTransport failed to connect to {127.0.0.1:2379 127.0.0.1 <nil> 0 <nil>}. Err: connection error: desc = "transport: Error while dialing parse \"127.0.0.1:2379\": first path segment in URL cannot contain colon". Reconnecting...
I0614 15:43:11.928672       9 clientconn.go:1114] [core] Subchannel Connectivity change to TRANSIENT_FAILURE
I0614 15:43:11.928688       9 balancer.go:207] [balancer] base.baseBalancer: handle SubConn state change: 0xc001306d20, TRANSIENT_FAILURE
I0614 15:43:15.989152       9 clientconn.go:435] [core] Channel Connectivity change to SHUTDOWN
I0614 15:43:15.989211       9 clientconn.go:1114] [core] Subchannel Connectivity change to SHUTDOWN
F0614 15:43:15.989229       9 instance.go:270] Error creating storage factory: context deadline exceeded
```

liggitt commented:
that looks suspicious:

> W0614 15:42:55.989219       9 clientconn.go:1326] [core] grpc: addrConn.createTransport failed to connect to {127.0.0.1:2379 127.0.0.1 <nil> 0 <nil>}. Err: connection error: desc = "transport: Error while dialing parse \"127.0.0.1:2379\": first path segment in URL cannot contain colon". Reconnecting...

liggitt commented:
this job passed successfully on this PR last week on https://github.com/kubernetes/kubernetes/commit/12d1a543d4f69cf34b94854fa99ce25647c51fb1 (see https://prow.k8s.io/pr-history/?org=kubernetes&repo=kubernetes&pr=100488)... changes since then:

* etcd server changed from 3.4.13 to 3.5.0-rc.0 (#102062)
* golang changed from go1.16.4 to go1.16.5 (#102689)
* konnectivity was enabled by default (#102661)

wongma7 commented:
it has something to do with the "egressDialer", the parse error seems to be coming from here https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go#L140 since gce tests set 
I0614 15:42:38.809712       9 flags.go:59] FLAG: --egress-selector-config-file="/etc/srv/kubernetes/egress_selector_configuration.yaml"

wongma7 commented:
I don't think it's PR `etcd preflight check was removed (#101993)` because even if the preflight check were in place, it would not have done anything. In the e2e tests etcd-servers is passed as "https://127.0.0.1:2379" so behavior is the same whether the check exists or not

wongma7 commented:
@liggitt I think it's because of this change in etcd 3.5 client https://github.com/etcd-io/etcd/commit/4a1c24556c22f55e75cb5167d555a8c367ad4d2b#diff-2bbaa3c83234888775aabf7697076e8035451045c479938e03ebb5c929132679L281-R281 . The egress dialer no longer gets passed a url with a scheme in it so its url.Parse errors.

liggitt committed:
Exercise egress configuration in local-up-cluster

liggitt committed:
Fix etcd egress dialer addr parsing

liggitt commented:
It was a change in the etcd client that previously passed a url to the dialer... that messed up our egress config parsing workaround. Reproduced simply by enabling egress proxy configuration with a "direct dialer" (see `Exercise egress configuration in local-up-cluster` commit) and fixed in `Fix etcd egress dialer addr parsing` commit

liggitt commented:
/retest

liggitt commented on go.mod:
```diff
@@ -227,24 +235,27 @@ replace (
 	github.com/go-bindata/go-bindata => github.com/go-bindata/go-bindata v3.1.1+incompatible
 	github.com/go-errors/errors => github.com/go-errors/errors v1.0.1
 	github.com/go-gl/glfw/v3.3/glfw => github.com/go-gl/glfw/v3.3/glfw v0.0.0-20200222043503-6f7a984d4dc4
+	github.com/go-kit/kit => github.com/go-kit/kit v0.9.0
```
@dims what does this pull in?

liggitt replied:
Looks like this was the version ahead of the go module update that increased the go module dependency tree (that landed in v0.10.0). Opened https://github.com/grpc-ecosystem/go-grpc-middleware/pull/426 to switch github.com/grpc-ecosystem/go-grpc-middleware to the slimmed down log module

liggitt commented:
@jeremyje @andyzhangx the aks-engine-windows-containerd failure looks unrelated to this PR and looks like a bug in the cluster config:

```
Method invocation failed because [System.Management.Automation.PSObject] does not contain a method named 'op_Addition'.
At C:\k\debug\collectlogs.ps1:111 char:25
+ ... pPorts | % { $tcpRangesArray += [pscustomobject]@{P1 = $_; P2 = $_} }
+                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (op_Addition:String) [], RuntimeException
    + FullyQualifiedErrorId : MethodNotFound
```

liggitt commented:
/retest

liggitt commented:
seeing dial timeout errors on storage-looking things both on pull-kubernetes-e2e-gce-storage-slow and pull-kubernetes-e2e-gce-ubuntu-containerd:

* https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100488/pull-kubernetes-e2e-gce-storage-slow/1404585852490747904/
* https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100488/pull-kubernetes-e2e-gce-ubuntu-containerd/1404585852406861824/

`error dialing backend: dial timeout`

@jsafrane @msau42 - is that a known issue on storage tests on master at the moment?

msau42 commented:
> seeing dial timeout errors on storage-looking things both on pull-kubernetes-e2e-gce-storage-slow and pull-kubernetes-e2e-gce-ubuntu-containerd:
> is that a known issue on storage tests on master at the moment?

I don't think so, our CI is showing a little flakiness, but not with `dial timeout`: https://k8s-testgrid.appspot.com/sig-storage-kubernetes#gce-slow

liggitt commented:
> > is that a known issue on storage tests on master at the moment?
> 
> I don't think so, our CI is showing a little flakiness, but not with `dial timeout`: https://k8s-testgrid.appspot.com/sig-storage-kubernetes#gce-slow

hmm... looks like it is pre-existing on master - https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&text=dial%20timeout&job=pull-kubernetes-e2e-gce-ubuntu-containerd, though maybe it ticked up a bit after konnectivity was enabled in this job on June 7 (cc @cheftako @jkh52)

liggitt commented:
/test pull-kubernetes-e2e-gce-storage-slow

liggitt commented:
/test pull-kubernetes-e2e-gce-storage-slow

Looks like this job is failing a lot (though not solidly) on other PRs too

liggitt commented:
This is ready for review, the pull-kubernetes-e2e-gce-storage-slow failures are spurious "dial timeout" errors and exist on master already

andyzhangx commented:
/retest
ignore `pull-kubernetes-e2e-aks-engine-windows-containerd` test failure since it failed to provision a cluster on Azure, failure is not related to this PR.

dims commented:
stats are a little bit bad, but expected with the fine grained go.mods now in etcd3 and the otel dependencies coming in transitively

```
2,4c2,4
< 	"totalDependencies": 364,
< 	"maxDepthOfDependencies": 32,
< 	"transitiveDependencies": 238
---
> 	"totalDependencies": 391,
> 	"maxDepthOfDependencies": 33,
> 	"transitiveDependencies": 263
```

dims commented:
only other dependency update of concern usually is `protobuf`. Hopefully we should be fine peeking at the diff:
https://github.com/golang/protobuf/compare/v1.4.3...v1.5.2

dims commented:
LGTM @liggitt will `/lgtm` after the rebase.

ptabor commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:
```diff
@@ -64,7 +67,7 @@ func init() {
 	utilruntime.Must(examplev1.AddToScheme(scheme))
 
-	capnslog.SetGlobalLogLevel(capnslog.CRITICAL)
+	clientv3.SetLogger(grpclog.NewLoggerV2(ioutil.Discard, ioutil.Discard, os.Stderr))
```
Let's use directly: 
  grpclog.SetLoggerV2(grpclog.NewLoggerV2(ioutil.Discard, ioutil.Discard, os.Stderr))

liggitt replied:
done in the later commit (Drop use of deprecated clientv3.SetLogger)

ptabor reviewed (approved):
Looks good to me. Thank you.

liggitt committed:
Update protobuf,grpc,etcd dependencies

1. Updated etcd/protobuf/grpc dependencies:

echo "
hack/pin-dependency.sh github.com/golang/protobuf latest
hack/pin-dependency.sh google.golang.org/protobuf latest
hack/pin-dependency.sh go.etcd.io/etcd/api/v3 v3.5.0-rc.0
hack/pin-dependency.sh go.etcd.io/etcd/client/v3 v3.5.0-rc.0
hack/pin-dependency.sh go.etcd.io/etcd/client/pkg/v3 v3.5.0-rc.0
hack/pin-dependency.sh go.etcd.io/etcd/pkg/v3 v3.5.0-rc.0
hack/pin-dependency.sh go.etcd.io/etcd/server/v3 v3.5.0-rc.0
hack/pin-dependency.sh go.etcd.io/etcd/tests/v3 v3.5.0-rc.0
hack/pin-dependency.sh google.golang.org/grpc latest
" | bash

2. Linted transitive dependencies until versions are clean:

hack/lint-dependencies.sh  | grep "  hack/pin-dependency.sh" | bash

3. Linted dependencies until dropped versions are clean:

hack/lint-dependencies.sh  | grep "dropreplace" | bash

4. Updated vendor and internal modules:

hack/update-vendor.sh
hack/update-internal-modules.sh

Repeated steps 2-4 until clean

liggitt commented:
rebased

dims commented:
/lgtm
/skip 

thanks @liggitt

liggitt commented:
/hold cancel

liggitt commented:
/skip
/test pull-kubernetes-e2e-gce-ubuntu-containerd

liggitt commented:
/skip

Jefftree commented:
LGTM for egress configs
